src/util.cc

```diff
@@ -423,21 +424,82 @@ const char* SpellcheckString(const char* text, ...) {
 }
 
 #ifdef _WIN32
+std::wstring Utf8ToWide(const std::string& u8_string) {
+  if (u8_string.empty())
+    return std::wstring();
+
+  size_t size_needed =
+      MultiByteToWideChar(CP_UTF8, 0, u8_string.data(),
+                          static_cast<int>(u8_string.size()), NULL, 0);
+  if (size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  std::wstring w_string(size_needed, 0);
+  int w_size_needed = MultiByteToWideChar(CP_UTF8, 0, u8_string.data(),
+                                          static_cast<int>(u8_string.size()),
+                                          &w_string[0], w_string.size());
+  if (w_size_needed == 0 || (w_size_needed > size_needed))
+    Fatal("Failed conversion to wide string");
+
+  return w_string;
+}
+
+std::string WideToUtf8(const std::wstring& w_string) {
+  if (w_string.empty())
+    return std::string();
+
+  int size_needed = WideCharToMultiByte(CP_UTF8, 0, w_string.data(),
+                                        static_cast<int>(w_string.size()), NULL,
+                                        0, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  std::string u8_string(size_needed, '\0');
+  int u8_size_needed = WideCharToMultiByte(
+      CP_UTF8, 0, &w_string[0], static_cast<int>(w_string.size()),
+      &u8_string[0], size_needed, NULL, NULL);
+  if (u8_size_needed == 0 || (u8_size_needed > size_needed))
+    Fatal("Failed conversion to UTF-8 string");
+
+  return u8_string;
+}
+
+char** convertCommandLine(int argc, wchar_t** wargv) {
+  std::wstring w_argString;
+  std::string argString;
+  char** argv = new char*[argc + 1];
+
+  for (int i = 0; i < argc; i++) {
+    // Convert the string
+    argString = WideToUtf8(wargv[i]);
+    // Prepare space in the vector
+    argv[i] = new char[argString.length() + 1];
+
+    // Copy the final results.
+    strcpy(argv[i], argString.c_str());
+  }
+
+  // Just to follow the c++ standard, we need to add this!
+  // Also, ninja will not work properly if we do not do so.
+  argv[argc] = NULL;
+  return argv;
+}
+
 string GetLastErrorString() {
   DWORD err = GetLastError();
 
-  char* msg_buf;
-  FormatMessageA(
+  wchar_t* msg_buf;
+  FormatMessageW(
         FORMAT_MESSAGE_ALLOCATE_BUFFER |
         FORMAT_MESSAGE_FROM_SYSTEM |
         FORMAT_MESSAGE_IGNORE_INSERTS,
         NULL,
         err,
         MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
-        (char*)&msg_buf,
+        (wchar_t*)&msg_buf,
```

Comment:
remove the cast.

Reply:
@jlonnberg

Reply:
Wait.

What the heck is happening here?

msg_buf is already wchar_t*, so we are taking the address of msg_buf (wchar_t**) and then casting it to wchar_t*?

How is this not causing stack corruption?

*sigh* oh my. This function is just ... disgusting.

https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessagew

Apparently this is the "right" way to use the function.

@jlonnberg 

Is there actually any benefit to using this version of the function, instead of the `FormatMessage()` version?

Reply:
Even though its ugly, I believe that it is the correct way. This way, if the message contains any wide chars, we get the correct conversion etc in the end.